mpqp/core/instruction/gates/native_gates.py

```diff
@@ -1092,6 +1157,79 @@ def __repr__(self) -> str:
         2
     )
 
+    def inverse(self) -> Gate:
+        return CRk_dagger(self.parameters[0], self.controls[0], self.targets[0])
+
+
+class CRk_dagger(RotationGate, ControlledGate):
+    """Two-qubit Controlled-Rk-dagger gate.
+
+    Args:
+        k: Parameter used in the definition of the phase to apply.
+        control: Index referring to the qubit used to control the gate.
+        target: Index referring to the qubit on which the gate will be applied.
+
+    Example:
+        >>> pprint(CRk_dagger(4, 0, 1).to_matrix())
+        [[1, 0, 0, 0],
+         [0, 1, 0, 0],
+         [0, 0, 1, 0],
+         [0, 0, 0, 0.9238795-0.3826834j]]
+
+    """
+
+    @classproperty
+    def braket_gate(cls):
+        from braket.circuits import gates
+
+        return gates.CPhaseShift
+
+    @classproperty
+    def qiskit_gate(cls):
+        from qiskit.circuit.library import CPhaseGate
+
+        return CPhaseGate
+
+    # TODO: this is a special case, see if it needs to be generalized
+    qlm_aqasm_keyword = "CNOT;PH"
+    qiskit_string = "cp"
+
+    def __init__(self, k: Expr | int, control: int, target: int):
+        self.parameters = [k]
+        ControlledGate.__init__(self, [control], [target], Rk_dagger(k, target), "CRk")
+        definition = UnitaryMatrix(
+            self.to_canonical_matrix(), **self.native_gate_options
+        )
+        ParametrizedGate.__init__(self, definition, [target], [k], "CRk")
+
+    @property
+    def theta(self) -> Expr | float:
+        r"""Value of the rotation angle, parametrized by ``k`` with the relation
+        `\theta = -\frac{\pi}{2^{k-1}}`."""
+        from sympy import pi
+
+        p = np.pi if isinstance(self.k, Integral) else pi
+        return -(p / 2 ** (self.k - 1))  # pyright: ignore[reportOperatorIssue]
+
+    @property
+    def k(self) -> Expr | float:
+        """See corresponding argument."""
+        return self.parameters[0]
+
+    def to_canonical_matrix(self):
+        e = exp(self.theta * 1j)  # pyright: ignore[reportOperatorIssue]
+        return np.array([[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 0, e]])
+
+    def __repr__(self) -> str:
+        return f"{type(self).__name__}({self.k}, {self.controls[0]}, {self.targets[0]})"
+
+    nb_qubits = (  # pyright: ignore[reportAssignmentType,reportIncompatibleMethodOverride]
+        2
+    )
+
+    def inverse(self) -> Gate:
+        return CRk(self.parameters[0], self.controls[0], self.targets[0])
```

Comment:
I would prefer `self.k` instead of `self.parameters[0]`: more readable and less susceptible to error in case we change how parameters are stored

Reply:
done